File: src/org/jgll/grammar/HeadGrammarSlot.java
package org.jgll.grammar;

import java.io.IOException;
import java.io.Writer;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

import org.jgll.grammar.slot.BodyGrammarSlot;
import org.jgll.grammar.slot.GrammarSlot;
import org.jgll.parser.GLLParserInternals;
import org.jgll.recognizer.GLLRecognizer;
import org.jgll.util.Input;

/**
 * 
 * The grammar slot corresponding to the head of a rule.
 * 
 * @author Ali Afroozeh
 * 
 */
public class HeadGrammarSlot extends GrammarSlot {
	
	private static final long serialVersionUID = 1L;

	private List<Alternate> alternates;
	
	private final Nonterminal nonterminal;
	
	private boolean nullable;
	
	private transient final Set<Terminal> firstSet;
	
	private transient final Set<Terminal> followSet;
	
	private transient final Set<HeadGrammarSlot> reachableNonterminals;
	
	public HeadGrammarSlot(Nonterminal nonterminal) {
		super(nonterminal.getName());
		this.nonterminal = nonterminal;
		this.alternates = new ArrayList<>();
		this.firstSet = new HashSet<>();
		this.followSet = new HashSet<>();
		this.reachableNonterminals = new HashSet<>();
	}
	
	public void addAlternate(Alternate alternate) {		
		alternates.add(alternate);
	}
	
	public void setAlternates(List<Alternate> alternates) {
		this.alternates = alternates;
	}
	
	public void removeAlternate(Alternate alternate) {
		alternates.remove(alternate);
	}
	
	public Set<Integer> getAlternateIndices() {
		Set<Integer> set = new HashSet<>();
		for(Alternate alternate : alternates) {
			set.add(alternate.getIndex());
		}
		return set;
	}
	
	public Set<Alternate> without(List<Symbol> list) {
		Set<Alternate> set = new HashSet<>(alternates);
		for(Alternate alternate : alternates) {
			if(alternate.match(list)) {
				set.remove(alternate);
				return set;
			}
		}
		return set;
	}
	
	public void remove(List<Symbol> list) {
		Iterator<Alternate> it = alternates.iterator();
		while(it.hasNext()) {
			Alternate alternate = it.next();
			if(alternate.match(list)) {
				it.remove();
			}
		}
	}
	
	public boolean isNullable() {
		return nullable;
	}
	
	public void setNullable(boolean nullable) {
		this.nullable = nullable;
	}
	
	@Override
	public GrammarSlot parse(GLLParserInternals parser, Input input) {
		for(Alternate alternate : alternates) {
			int ci = parser.getCurrentInputIndex();
			BodyGrammarSlot slot = alternate.getFirstSlot();
			if(slot.testFirstSet(ci, input) || (slot.isNullable() && slot.testFollowSet(ci, input))) {
				parser.addDescriptor(slot);
			}
		}
		return null;
	}
	
	@Override
	public GrammarSlot recognize(GLLRecognizer recognizer, Input input) {
		for(Alternate alternate : alternates) {
			int ci = recognizer.getCi();
			BodyGrammarSlot slot = alternate.getFirstSlot();
			if(slot.testFirstSet(ci, input) || (slot.isNullable() && slot.testFollowSet(ci, input))) {
				org.jgll.recognizer.GSSNode cu = recognizer.getCu();
				recognizer.add(alternate.getFirstSlot(), cu, ci);
			}
		}
		return null;
	}
	
	@Override
	public void codeParser(Writer writer) throws IOException {
		writer.append("// " + nonterminal.getName() + "\n");
		writer.append("private void parse_" + id + "() {\n");
		for (Alternate alternate : alternates) {
			writer.append("   //" + alternate.getFirstSlot() + "\n");
			alternate.getFirstSlot().codeIfTestSetCheck(writer);			
			writer.append("   add(grammar.getGrammarSlot(" + alternate.getFirstSlot().getId() + "), cu, ci, DummyNode.getInstance());\n");
			writer.append("}\n");
		}
		writer.append("   label = L0;\n");
		writer.append("}\n");

		for (Alternate alternate : alternates) {
			writer.append("// " + alternate + "\n");
			writer.append("private void parse_" + alternate.getFirstSlot().getId() + "() {\n");
			alternate.getFirstSlot().codeParser(writer);
		}
	}
	
	public Alternate getAlternateAt(int index) {
		return alternates.get(index);
	}
	
	public List<Alternate> getAlternatesIncludingNull() {
		return alternates;
	}

	public List<Alternate> getAlternates() {
		return new ArrayList<>(alternates);
	}
	
	public Nonterminal getNonterminal() {
		return nonterminal;
	}
		
	public Set<Terminal> getFirstSet() {
		return firstSet;
	}
	
	public Set<Terminal> getFollowSet() {
		return followSet;
	}
	
	public int getCountAlternates() {
		return alternates.size();
	}
	
	public Set<HeadGrammarSlot> getReachableNonterminals() {
		return reachableNonterminals;
	}
	
	public boolean contains(List<Symbol> list) {
		for(Alternate alternate : alternates) {
			if(alternate.match(list)) {
				return true;
			}
		}
		return false;
	}

	@Override
	public String toString() {
		return nonterminal.toString();
	}

}

File: src-scala/org/jgll/grammar/slot/HeadGrammarSlot.scala
package org.jgll.grammar.slot

import java.io.Writer
import org.jgll.grammar._
import org.jgll.parser.GLLParserInternalsTrait
import org.jgll.recognizer.GLLRecognizerTrait
import scala.reflect.{BeanProperty, BooleanBeanProperty}

import scala.collection.mutable._
import org.jgll.util.InputTrait

trait HeadGrammarSlotTrait {
  self: GrammarSlotTrait
   with AlternateTrait
   with GLLParserInternalsTrait
   with InputTrait
   with TerminalTrait
   with GLLRecognizerTrait
   =>
  @SerialVersionUID(1L)
  class HeadGrammarSlot(@BeanProperty val nonterminal: Nonterminal) extends GrammarSlot {

    private var alternates: ListBuffer[Alternate] = ListBuffer[Alternate]()

    @BooleanBeanProperty
    var nullable: Boolean = _

    @BooleanBeanProperty
    var directNullable: Boolean = _

    @BeanProperty
    val firstSet = HashSet[Terminal]()

    @BeanProperty
    val followSet = HashSet[Terminal]()

    @BeanProperty
    var epsilonAlternate: Alternate = _

    def addAlternate(alternate: Alternate) {
      alternates += (alternate)
    }

    def setAlternates(alternates: ListBuffer[Alternate]) {
      this.alternates = alternates
    }

    def removeAlternate(alternate: Alternate) {
      alternates -= (alternate)
    }

    def without(list: ListBuffer[Symbol]): Set[Alternate] = {
      val set = new HashSet[Alternate]() ++= alternates
      for (alternate <- alternates if alternate.`match`(list)) {
        set.remove(alternate)
      }
      set
    }

    def without(withoutSet: Set[ListBuffer[Symbol]]): Set[Alternate] = {
      val set = new HashSet[Alternate]() ++= alternates
      for (alternate <- alternates; list <- withoutSet if alternate.`match`(list)) {
        set.remove(alternate)
      }
      set
    }

    def remove(list: ListBuffer[Symbol]) {
      alternates foreach { alternate =>
        if (alternate.`match`(list))
          alternates -= alternate
      }
    }

    def removeAllAlternates() {
      alternates.clear()
    }

    def setNullable(nullable: Boolean, directNullable: Boolean) {
      this.nullable = nullable
      this.directNullable = directNullable
    }

    override def parse(parser: GLLParserInternals, input: Input): GrammarSlot = {
      for (alternate <- alternates) {
        val ci = parser.getCurrentInputIndex
        val slot = alternate.getFirstSlot
        if (slot.testFirstSet(ci, input) || (slot.isNullable && slot.testFollowSet(ci, input))) {
          parser.addDescriptor(slot)
        }
      }
      null
    }

    override def recognize(recognizer: GLLRecognizer, input: Input): GrammarSlot = {
      for (alternate <- alternates) {
        val ci = recognizer.getCi
        val slot = alternate.getFirstSlot
        if (slot.testFirstSet(ci, input) || (slot.isNullable && slot.testFollowSet(ci, input))) {
          val cu = recognizer.getCu
          recognizer.add(alternate.getFirstSlot, cu, ci)
        }
      }
      null
    }

    override def codeParser(writer: Writer) {
      writer.append("// " + nonterminal.getName + "\n")
      writer.append("private void parse_" + id + "() {\n")
      for (alternate <- alternates) {
        writer.append("   //" + alternate.getFirstSlot + "\n")
        alternate.getFirstSlot.codeIfTestSetCheck(writer)
        writer.append("   add(grammar.getGrammarSlot(" + alternate.getFirstSlot.getId + "), cu, ci, DummyNode());\n")
        writer.append("}\n")
      }
      writer.append("   label = L0;\n")
      writer.append("}\n")
      for (alternate <- alternates) {
        writer.append("// " + alternate + "\n")
        writer.append("private void parse_" + alternate.getFirstSlot.getId +
          "() {\n")
        alternate.getFirstSlot.codeParser(writer)
      }
    }

    def getAlternateAt(index: Int): Alternate = alternates(index)

    def getAlternates(): ListBuffer[Alternate] = ListBuffer[Alternate]() ++= alternates

    def getAlternatesAsSet(): Set[Alternate] = new HashSet() ++= alternates

    def getCountAlternates(): Int = alternates.size

    def contains(list: ListBuffer[Symbol]): Boolean = {
      alternates.find(_.`match`(list)).isDefined
    }

    def contains(set: Set[ListBuffer[Symbol]]): Boolean = {
      set.find(contains(_)).isDefined
    }

    override def toString(): String = nonterminal.toString
  }
}